Fix code for v0.19.0

Vindaar opened the pull request:
First of all, this PR is based on @alehander42 PR to fix the `nil` strings.

Besides this, there's been several changes which also cause regressions in the test cases. From the top of my head:
- null terminator in strings not accessible anymore
- compile time behavior of `when` expressions is forced a bool, but `quote do` returns an int

And as it seems converting an integer, which doesn't fit into some int type throws a different error? The case that tries it fails, so I hacked that.

With these changes the test suite works up to `Load tuple - missing field`. There it  still crashes. I'll try to see what's going on there/

Maybe (some) of the changes I made are stupid ways to handle the stuff? Quite possibly, since I don't know the code base. The hardest part for me is the internal exception handling, since I can't see where the ACTUAL exception is being raised, since it's replaced by the "user focused" stack trace, which simply points to the YAML call in the *user* file and a note "please report the bug".

Vindaar commented on test/tserialization.nim:
```diff
@@ -205,16 +205,16 @@ suite "Serialization":
+  #test "Dump nil string":
+  #  let input: string = ""
+  #  var output = dump(input, tsNone, asTidy, blockOnly)
+  #  assertStringEqual yamlDirs & "\n!n!nil:string \"\"", output
```
Yeah, you're right. But that's why this PR is still called WIP. :P

Vindaar commented:
Thanks! I'll fix your requested changes tomorrow. I mainly still need to figure out why some tests are still broken though. :)

Vindaar commented on yaml/serialization.nim:
```diff
@@ -1163,10 +1169,11 @@ proc constructChild*[T](s: var YamlStream, c: ConstructionContext,
-      result = nil
-      return
+    #if item.scalarTag == yTagNimNilSeq:
+    # TODO: this whole branch still makes sense
```
That comment of mine was a question to myself, hehe. Thanks for the explanation!

Vindaar commented on test/tserialization.nim:
```diff
@@ -97,7 +97,7 @@ proc constructObject*(s: var YamlStream, c: ConstructionContext,
 
 template assertStringEqual(expected, actual: string) =
-  for i in countup(0, min(expected.len, actual.len)):
+  for i in countup(0, min(expected.len, actual.len) - 1):
```
I replaced this by a more complex version, which tries to check 
- if the string content up to the minimal length is the same
- if strings are of different length shows the additional part that isn't in the other.

I added a test case for your example.

Vindaar commented on yaml/parser.nim:
```diff
@@ -158,7 +158,7 @@ proc currentScalar(c: ParserContext, e: var YamlStreamEvent)
                       scalarAnchor: c.anchor)
   shallowCopy(e.scalarContent, c.lex.buf)
-  c.lex.buf = cast[string not nil](newStringOfCap(256))
+  c.lex.buf = cast[string](newStringOfCap(256))
```
Yeah, good point.

Vindaar commented on yaml/serialization.nim:
```diff
@@ -199,7 +199,13 @@ proc constructObject*[T: int8|int16|int32|int64](
+        result = T(nInt)
+      else:
+        # if outside of range, what to do?!
+        raise newException(YamlConstructionError, "AAA")
```
Done.

Vindaar commented:
On my machine there's only a single test case, which fails. Will fix that too.

I also had to fix one proc in the lexer, which also depended on accessing the null terminator of a string. The solution is pretty verbose and maybe you can come up with a nicer (more scalable) solution for that?

And I'll also address the last two remaining comments of yours ~tomorrow.

Vindaar commented:
@flyx Ok, so I've addressed all your comments. Two questions:
1. I'm not sure if my extension of `Level` in tojson is quite correct. I'm resetting the `expKey` field of the tuple to `true`, if the `key` is being set to `""`, which I assume is not correct. I guess setting an empty key shouldn't always indicate that a key is expected afterwards. 
2. with the changes now all tests but [S98Z] pass. The expected output of that is the empty string, but we recover the `\n`and the comment. I guess I introduced that regression due to the changes I made to deal with the null terminator not being accessible. My understanding of yaml and this library isn't quite suited to debug this. Any hint on where I should start?

flyx commented on test/tserialization.nim:
```diff
@@ -124,6 +142,19 @@ proc newNode(v: string): ref Node =
 let blockOnly = defineOptions(style=psBlockOnly)
 
 suite "Serialization":
+  test "Internal string assert":
```
I get that you tested your code with this, but I'd rather not have tests testing the test code, even less tests that produce random files. Please remove this and the `file` paramater of `assertStringEquals`.

Vindaar replied:
Ok, up to you!

flyx commented on yaml/private/internal.nim:
```diff
@@ -5,16 +5,24 @@
-      try: writeStackTrace()
+      try:
+        writeStackTrace()
+        when defined(debug):
```
I think this is not necessary given that we are already in `when not defined(release)`. Also, shouldn't you check whether `exc.parent` is `nil`?

Vindaar replied:
Ugh, didn't event think of checking release... Good point, also regarding checking for nil!

flyx commented on yaml/private/internal.nim:
```diff
@@ -23,7 +31,12 @@ template yAssert*(e: typed) =
-        try: writeStackTrace()
+        try:
+          writeStackTrace()
+          when defined(debug):
```
See above.

flyx commented on yaml/private/lex.nim:
```diff
@@ -187,7 +190,10 @@ proc at(lex: YamlLexer, t: typedesc[BaseLexer], pos: int): char {.inline.} =
+  if lex.sSource.src.len == pos:
+    '\0'
+  else:
+    lex.sSource.src[pos]
```
Well that's one way to fix it. However, it makes these checks necessary everywhere. A far simpler (and more performant) solution would be to append `'\0'` to the string when creating a `StringSource`.

Vindaar replied:
Ahh, thanks. That's what I would have done, if I understood where I'd have to do that, hehe. I'll check that.

Vindaar commented:
Yup, thanks. I also prefer to not have to change the lexer. I just don't understand the code enough to know where that change is necessary. I'll take a look at `StringSource` creation tomorrow and change the lexer back! :)
Any idea regarding the broken test case?

Vindaar commented:
Ok, so I addressed all your comments. Lexer is reverted back to its original form and instead we just append a null terminator to the source. 

One thing about that though: at first I only added a null terminator to the `StringSource.src`. But to avoid having to duplicate the same in the JS branch I instead defined a `sourceNull` at the beginning of the `newYamlLexer` proc. That should mean we copy the source before assigning it to the `StringSource` though (needed for JS as far as I understand, strings being immutable there, but unnecessary for C). Is that ok?

flyx commented on yaml/private/internal.nim:
```diff
@@ -5,16 +5,24 @@
-      try: writeStackTrace()
+      try:
+        writeStackTrace()
+        echo "Internal stacktrace:"
```
Okay, last change I request: This `echo` should be inside the `if` branch below. Doesn't make sense without actually having a parent stacktrace displayed.

Vindaar replied:
Yeah, good point.

flyx commented on yaml/private/internal.nim:
```diff
@@ -23,7 +31,12 @@ template yAssert*(e: typed) =
-        try: writeStackTrace()
+        try:
+          writeStackTrace()
+          echo "Internal stacktrace:"
```
Same here.

Vindaar commented:
Done. :)
